File: java/src/com/zhihu/rucene/Benchmark.java
package com.zhihu.rucene;

import java.nio.file.Paths;
import java.nio.file.Files;
import java.io.*;
import java.util.*;
import java.util.concurrent.atomic.*;
import java.nio.charset.StandardCharsets;
import java.nio.charset.*;

import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.analysis.core.WhitespaceAnalyzer;
import org.apache.lucene.document.Document;
import org.apache.lucene.index.*;
import org.apache.lucene.queryparser.classic.QueryParser;
import org.apache.lucene.search.*;
import org.apache.lucene.store.*;
import org.apache.lucene.util.*;

public class Benchmark {
  public static class BenchmarkResult {
    long count;
    String query;
    long elapsed;

    BenchmarkResult(long count, long elapsed, String query) {
      this.elapsed = elapsed;
      this.count = count;
      this.query = query;
    }
  }

  public static void run(IndexSearcher searcher, String field, String queriesInput, int limit) throws IOException {
    Analyzer analyzer = new WhitespaceAnalyzer();
    BufferedReader queries = Files.newBufferedReader(Paths.get(queriesInput), StandardCharsets.UTF_8);
    QueryParser parser = new QueryParser(field, analyzer);

    /* execute queries */
    String query;
    ArrayList<BenchmarkResult> results = new ArrayList();
    long start = System.currentTimeMillis();
    while ((query = queries.readLine()) != null) {
      query = query.trim();
      if (!query.isEmpty()) {
        try {
          query = query.trim();
          Query realQuery = parser.parse(query);
          long queryStart = System.currentTimeMillis();
          TopDocs docs = searcher.search(realQuery, limit);
          results.add(new BenchmarkResult(docs.scoreDocs.length, System.currentTimeMillis() - queryStart, query));
        } catch (Exception ignored) {
        }
      }
    }
    long elapsed = System.currentTimeMillis() - start;
    Collections.sort(results, (o1, o2) -> {
      if (o1.elapsed < o2.elapsed) {
        return -1;
      } else if (o1.elapsed > o2.elapsed) {
        return 1;
      } else {
        return 0;
      }
    });

    int len = results.size();
    System.out.println("======== Benchmark with early terminating of " + limit + "========");
    System.out.println(results.size() + " queries executed in " + elapsed + " ms, " + (elapsed / len) + " ms per query on average");
    BenchmarkResult result = results.get(len / 2);
    System.out.println("p50: '" + result.query + "' returned " + result.count + " results in " + result.elapsed + " ms");
    result = results.get(len * 75 / 100);
    System.out.println("p75: '" + result.query + "' returned " + result.count + " results in " + result.elapsed + " ms");
    result = results.get(len * 90 / 100);
    System.out.println("p90: '" + result.query + "' returned " + result.count + " results in " + result.elapsed + " ms");
    result = results.get(len * 95 / 100);
    System.out.println("p95: '" + result.query + "' returned " + result.count + " results in " + result.elapsed + " ms");
    result = results.get(len * 99 / 100);
    System.out.println("p99: '" + result.query + "' returned " + result.count + " results in " + result.elapsed + " ms");
    result = results.get(len * 999 / 1000);
    System.out.println("p999: '" + result.query + "' returned " + result.count + " results in " + result.elapsed + " ms");
    result = results.get(len - 1);
    System.out.println("max: '" + result.query + "' returned " + result.count + " results in " + result.elapsed + " ms");
  }

  public static void main(String[] args) throws Exception {
    /* index path, field, search file, rucene path */
    if (args.length != 3) {
      throw new RuntimeException("missing required arguments");
    }
    IndexReader reader = DirectoryReader.open(FSDirectory.open(Paths.get(args[0])));

    IndexSearcher searcher = new IndexSearcher(reader);
    run(searcher, args[1], args[2], 100);
    run(searcher, args[1], args[2], 50);
    run(searcher, args[1], args[2], 100);
  }
}

File: src/bin/benchmark.rs
extern crate rucene;

use std::env;
use std::fs::File;
use std::io::*;
use std::path::Path;
use std::sync::Arc;
use std::time;
use std::usize;

use rucene::core::index::StandardDirectoryReader;
use rucene::core::search::collector::{EarlyTerminatingSortingCollector, TopDocsCollector};
use rucene::core::search::query_string::*;
use rucene::core::search::searcher::*;
use rucene::core::store::MmapDirectory;

type QueryResult = (String, usize, time::Duration);

fn benchmark_queries(
    searcher: &IndexSearcher,
    field: &str,
    queries: &[String],
    limit: usize,
) -> Vec<QueryResult> {
    let mut results = Vec::new();
    for query in queries {
        let start = time::Instant::now();
        let mut top_collector = TopDocsCollector::new(100);
        let real_query =
            QueryStringQueryBuilder::new(query.clone(), vec![(field.into(), 1.0)], 1, 1.0)
                .build()
                .unwrap();
        if limit != usize::MAX {
            let mut early_collector =
                EarlyTerminatingSortingCollector::new(&mut top_collector, limit);
            searcher
                .search(real_query.as_ref(), &mut early_collector)
                .unwrap();
        } else {
            searcher
                .search(real_query.as_ref(), &mut top_collector)
                .unwrap();
        }
        results.push((
            query.clone(),
            top_collector.top_docs().score_docs().len(),
            start.elapsed(),
        ));
    }

    results.sort_unstable_by_key(|r| r.2);

    return results;
}

fn format_elapsed(duration: &time::Duration) -> String {
    let secs = duration.as_secs();
    let nanos = duration.subsec_nanos();
    let ms = nanos / 1000000;
    let us = (nanos % 1000000) / 1000;
    let ns = nanos % 1000;
    if secs > 0 {
        format!("{}.{} secs", secs, ms)
    } else if ms > 0 {
        format!("{}.{} ms", ms, us)
    } else {
        format!("{}.{} us", us, ns)
    }
}

fn format_result(result: &QueryResult) -> String {
    format!(
        "'{}' took {} returned {} docs",
        result.0,
        format_elapsed(&result.2),
        result.1
    )
}

fn benchmark_with_limit(searcher: &IndexSearcher, field: &str, queries: &[String], limit: usize) {
    if limit == usize::MAX {
        println!("======== Benchmark without early terminating ========");
    } else {
        println!(
            "======== Benchmark with early terminating of {} ========",
            limit
        );
    }
    let start = time::Instant::now();
    let results = benchmark_queries(searcher, field, queries, limit);
    let len = results.len();
    println!(
        "{} queries executed in {}, {} per query on average",
        queries.len(),
        format_elapsed(&start.elapsed()),
        format_elapsed(&(start.elapsed() / queries.len() as u32))
    );
    println!("p50: {}", format_result(&results[len / 2]));
    println!("p75: {}", format_result(&results[len * 75 / 100]));
    println!("p90: {}", format_result(&results[len * 90 / 100]));
    println!("p95: {}", format_result(&results[len * 95 / 100]));
    println!("p99: {}", format_result(&results[len * 99 / 100]));
    println!("p999: {}", format_result(&results[len * 999 / 1000]));
    println!("max: {}", format_result(&results[len - 1]));
}

fn benchmark(index: String, field: String, input: String) {
    let directory = Arc::new(MmapDirectory::new(&Path::new(&index), 1024 * 1024).unwrap());
    let queries: Vec<String> = File::open(&input)
        .map(BufReader::new)
        .map(BufReader::lines)
        .map(|i| i.filter_map(Result::ok))
        .map(Iterator::collect)
        .expect("No query to run");

    let reader = Arc::new(StandardDirectoryReader::open(directory).unwrap());
    let searcher = Arc::new(IndexSearcher::new(reader.clone()));
    benchmark_with_limit(searcher.as_ref(), &field, &queries, 100);
    benchmark_with_limit(searcher.as_ref(), &field, &queries, 50);
    benchmark_with_limit(searcher.as_ref(), &field, &queries, 100);
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 4 {
        panic!("Missing required arguments");
    }
    benchmark(args[1].clone(), args[2].clone(), args[3].clone());
}
